Add a constituency model for Icelandic

ingunnjk opened the pull request:
**BEFORE YOU START**: please make sure your pull request is against the `dev` branch. 
We cannot accept pull requests against the `main` branch. 
See our [contributing guide](https://github.com/stanfordnlp/stanza/blob/main/CONTRIBUTING.md) for details.

## Description
I want to add an Icelandic constituency model for Icelandic. I have made changes to prepare_con_dataset.py and added the script convert_icepahc.py in stanza/utils/datasets/constituency. I have also added an Icelandic BERT model to stanza/resources/default_packages.py.

## Fixes Issues
A list of issues/bugs with # references. (e.g., #123)

## Unit test coverage
Are there unit tests in place to make sure your code is functioning correctly?
(see [here](https://github.com/stanfordnlp/stanza/blob/master/tests/test_tagger.py) for a simple example)

## Known breaking changes/behaviors
None.

AngledLuffa commented:
Hey, thanks for sending this.  Very helpful.

Are you familiar with squashing?  If so, would you be up for squashing commit #1 with #2 and #3 with #4?  If not, I can do it on my end.

In your change comments, you talk about how there is a version which is simplified already.  Is it necessary to use that for the parser?  If yes, did you use a script to do those simplifications?

ingunnjk commented:
Hi! 

I have now squashed the commits! 

I tried using the original IcePaHC file for the parser but that didn't seem to work. I also removed some things that shouldn't be a part of the parsing like empty nodes (traces, comments etc.) and some of it was by code and some by hand so there is unfortunately not a script for that, I'm sorry.

Also one other thing, you added * to the list of functional tags to cut off since * is used as a separator in the treebank file that is used for the training. I would actually prefer if it is not cut off! I removed the ending of most constituents, like ADJP*OC, but I would like to keep some of them that I think would be useful, for example IP*MAT. I trained a model where * was cut off and then also when it is not cut off and the difference in scoring is only like 0,25% (from 90,63% to 90,38%). If it would then be possible to change the * to - in the output from the parser, that would be greatly appreciated (or change * to - beforehand and not cutting - off for the Icelandic parser?).

AngledLuffa commented:
> I also removed some things that shouldn't be a part of the parsing like empty nodes (traces, comments etc.) and some of it was by code and some by hand so there is unfortunately not a script for that, I'm sorry.

The tree reading code automatically removes empty nodes.  Perhaps it would be possible to recreate the conversion by hand.  I'm not too opposed to using editing files, but it does look like they update this corpus fairly often:

https://github.com/antonkarl/icecorpus

which suggests that being able to automatically convert their work to the internal Stanza format would be useful.

> Also one other thing, you added `*` to the list of functional tags to cut off since `*` is used as a separator in the treebank file that is used for the training

I see how that could be useful, in that English also has some tags such as `NP-TMP` which could be kept instead of discarded.  I'll have to think about how to do that.  Probably pass in the dataset as a parameter to `read_treebank`

ingunnjk commented:
Okay! So would I have to do (or rather update) the conversion script in order to make the Icelandic constituency model a part of Stanza or could that come as an update later on?

AngledLuffa commented:
In my ideal world, it would show up all at once :)  If that doesn't work with you, we can also start from your converted version of the treebank.  If you can send us the partially finished script, I can also help as needed, since you're already donating a bunch of work to us.

AngledLuffa commented:
Any room for me to help with the conversion script?

ingunnjk commented:
Sorry for the late answer, I haven't been able to work on this for the past weeks. In the conversion script (since the tree reading code automatically removes empty nodes) it is mostly a matter of removing lemmas (so _(NS-D löndum-land)_ becomes _(NS-D löndum)_) and combining word parts that start/end with $ (in most cases a noun and its determiner), for example should _(NS-N tungur$-tunga) (D-N <span>$</span>nar-hinn)_ become _(NS-N tungurnar)_ and _(RP upp<span>$</span>-upp) (VBPI $heldur-halda)_ should become _(VBPI uppheldur)_. I don't know exactly when I'll be able to continue working on this (hopefully next week) but if you have the time you could for sure help move it along.

AngledLuffa commented:
I checked in a script which uses Tsurgeon to convert a single tree:

https://github.com/stanfordnlp/stanza/commit/6b468b96e321431fbe894c24b907a2f8af62dca5

If you have a list of cases where `$` combines two words into one word, I can expand that to cover the rest of the treebank.  Obviously it's easy enough to combine neighboring `$` words, but I'm not sure which ones take precedence in cases other than the two you've mentioned so far.

I put a brief introduction using this example on the github.io

https://stanfordnlp.github.io/stanza/tsurgeon.html

It doesn't really say anything other than the example I just checked in, though.

PS sorry for reusing your script's name and therefore making it slightly more annoying to do this PR

AngledLuffa commented:
Any thoughts on the tregex / tsurgeon solution for this issue?

AngledLuffa commented:
@ingunnjk any thoughts on how best to proceed with the conversion of the Icelandic dataset?  If you have a set of heuristics on how to combine the `$` nodes, I can turn that into a sequence of Tsurgeon expressions which hopefully handles the conversion from the raw dataset

AngledLuffa commented:
Can I follow up on this?  Happy to add the model, but need some insight on adapting the conversion.  Looking to make a new release in the next couple weeks

ingunnjk commented:
Hi, very sorry for the late answer! I wont be able to work on this in November but I will have time in the beginning of December. I will do my very best to finish this then, it would be great to be able to add the model.

From: John Bauer ***@***.***>
Date: Thursday, 24 October 2024 at 07:39
To: stanfordnlp/stanza ***@***.***>
Cc: Ingunn Jóhanna Kristjánsdóttir - HI ***@***.***>, Mention ***@***.***>
Subject: Re: [stanfordnlp/stanza] Add a constituency model for Icelandic (PR #1389)

Can I follow up on this? Happy to add the model, but need some insight on adapting the conversion. Looking to make a new release in the next couple weeks

—
Reply to this email directly, view it on GitHub<https://github.com/stanfordnlp/stanza/pull/1389#issuecomment-2434525791>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ATGDTHIJ4FCX2ZEL5VDB4CLZ5CPZ5AVCNFSM6AAAAABGXGSW26VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMZUGUZDKNZZGE>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

AngledLuffa commented:
No worries.  Now-ish or in the next couple weeks would be a perfect time, as I'm rebuilding the models for a new version of Stanza using a new version of UD.  Later works too.  If you can summarize some of the tree edits needed to get from the original treebank to the version used for training, I can help redo those in tregex & tsurgeon and also rebuild the models if needed.  Thanks!
